src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -269,30 +288,83 @@ Pipe MergeTreeDataSelectExecutor::readFromParts(
 
     /// Select the parts in which there can be data that satisfy `minmax_idx_condition` and that match the condition on `_part`,
     ///  as well as `max_block_number_to_read`.
+    /// Skip parts fingerprints if any to the query context, or skip parts which fingerprints marked as excluded.
     {
-        auto prev_parts = parts;
-        parts.clear();
 
-        for (const auto & part : prev_parts)
-        {
-            if (part_values.find(part->name) == part_values.end())
-                continue;
+        Context & query_context = const_cast<Context &>(context).getQueryContext();
 
-            if (part->isEmpty())
-                continue;
-
-            if (minmax_idx_condition && !minmax_idx_condition->checkInHyperrectangle(
-                    part->minmax_idx.hyperrectangle, data.minmax_idx_column_types).can_be_true)
-                continue;
+        auto process_parts = [&]() {
+            auto excluded_fingerprints_map = query_context.getExcludedFingerprints()->getMap();
+            std::set<String> temp_fingerprints;
+            auto prev_parts = parts;
+            parts.clear();
 
-            if (max_block_numbers_to_read)
+            for (const auto & part : prev_parts)
             {
-                auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
-                if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                if (part_values.find(part->name) == part_values.end())
+                    continue;
+
+                if (part->isEmpty())
                     continue;
+
+                if (minmax_idx_condition
+                    && !minmax_idx_condition->checkInHyperrectangle(part->minmax_idx.hyperrectangle, data.minmax_idx_column_types).can_be_true)
+                    continue;
+
+                if (max_block_numbers_to_read)
+                {
+                    auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
+                    if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                        continue;
+                }
+
+                if unlikely(!part->fingerprint.empty())
+                {
+                    /// Skip a part if its fingerprint is meant to be excluded
+                    if unlikely (excluded_fingerprints_map.find(part->fingerprint) != excluded_fingerprints_map.end())
+                        continue;
+
+                    auto result = temp_fingerprints.insert(part->fingerprint);
+                    // found a block with the same fingerprint one the same replica, ignore it
+                    if (!result.second)
```

Comment:
How is it possible? Only because of manual error?

Reply:
>Only because of manual error?
correct

as we haven't yet decided on how to generate fingerprints, there's a chance to make a mistake copy part on a local replica.
can be removed if you think it should be handled somewhere else.